lib/wlib/memory/Memory.h

```diff
@@ -53,85 +55,172 @@ static MemoryInitDestroy __g_smemoryInitDestroy;
+ * only supports default Ctors for objects
+ *
+ * @pre malloc creates number of blocks of given type and it does not allocate memory by providing it
+ * number of bytes
+ *
+ * @tparam T pointer array type
+ * @tparam NonArrayType pointer non array type
+ * @param num number of elements needed
+ * @return address to memory allocated
+ */
+template<
+        typename T,
+        typename NonArrayType = typename wlp::remove_extent<T>::type,
+        typename = typename wlp::enable_if<wlp::is_array<T>::value, bool>::type>
+NonArrayType *malloc(wlp::size_type num){
+    if (wlp::is_fundamental<typename wlp::remove_extent<T>::type>::value){
+        void *memory = __memory_alloc(static_cast<wlp::size32_type>(sizeof(NonArrayType)) * num, false);
+        return static_cast<NonArrayType *>(memory);
     }
 
-    memory = __memory_alloc(static_cast<wlp::size32_type>(sizeof(Type)) * num, false);
+    void *memory = __memory_alloc(static_cast<wlp::size32_type>(sizeof(NonArrayType)) * num, true);
 
-    return static_cast<Type *>(memory);
-}
+    uint16_t *objInfo = static_cast<uint16_t *>(memory);
+    *objInfo = static_cast<uint16_t>(num);
+
+    char *pointer = reinterpret_cast<char *>(++objInfo);
+    for (wlp::size_type i = 0; i < num; ++i) {
+        new(pointer) NonArrayType();
+        pointer += sizeof(NonArrayType);
+    }
+
+    return reinterpret_cast<NonArrayType *>(objInfo);
+};
 
 /**
- * This reallocates the memory to accommodate the new size provided. The memory address has to be
- * allocation from Memory otherwise results are undefined. If there is not a sufficient memory
+ * This allocates memory of the size of the type provided. Memory allocated could be greater
+ * than what has been asked in order to accommodate fixed memory allocations. If there is not a sufficient memory
  * available, it returns a @code nullptr @endcode
  *
- * @pre realloc creates number of blocks of given type and it does not allocate memory by proving it
+ * This is only used for creating multiple block of memory and hence should be used for arrays. This
+ * is a special case of @code malloc @endcode which supports creation of an array of objects where you can
+ * use non default ctor
+ *
+ * @pre this creates number of blocks of given type and it does not allocate memory by providing it
  * number of bytes
  *
+ * @warning should be avoided and used only where this is must
+ *
+ * @tparam T pointer array type
+ * @tparam Args Argument types for non-fundamental Types
+ * @param args the arguments for @code Args @end types
+ * @tparam NonArrayType pointer non array type
+ * @param num number of elements needed
+ * @return address to memory allocated
+ */
+template<
+        typename T,
+        typename... Args,
+        typename NonArrayType = typename wlp::remove_extent<T>::type,
+        typename = typename wlp::enable_if<wlp::is_fundamental<NonArrayType>::value>,

```

Comment:
If I'm reading this correctly you have it has enabled only if the non array type is **not** an object?

Reply:
Wait I have tested this with an object in `memory_check` file and it worked lol. But yeah this line we should take it off